graphix/channels.py

```diff
@@ -2,78 +2,137 @@
+        except ValueError as e:
+            raise ValueError("Failed to intialize KrausChannel object.") from e
+
+        if len(nqubits) != 1:
+            raise ValueError("All operators must have the same shape.")
+        (nqubit,) = nqubits
 
-    def __init__(self, kraus_data):
+        return nqubit
+
+    def __init__(self, kraus_data: Iterable[KrausData]) -> None:
         """Initialize `KrausChannel` given a Kraus operator.
 
         Parameters
         ----------
-        kraus_data : list
-            array of Kraus operator data.
-            list(dict): [{coef: scalar, operator: list}, {parameter: scalar, operator: list}, ...]
-            only works for square Kraus operators
+        kraus_data : Iterable[KrausData]
+            Iterable of Kraus operator data.
 
         Raises
         ------
         ValueError
-            If empty list is provided.
+            If kraus_data is empty.
         """
-        # check there is data
         if not kraus_data:
             raise ValueError("Cannot instantiate the channel with empty data.")
 
-        if not isinstance(kraus_data, (list, np.ndarray, tuple)):
-            raise TypeError(f"The data must be a list, a numpy.ndarray or a tuple not a {type(kraus_data)}.")
+        self.__nqubit = self._nqubit(kraus_data)
+        self.__data = list(copy.deepcopy(data) for data in kraus_data)
 
-        # check that data is correctly formatted before assigning it to the object.
-        assert check_data_values_type(kraus_data)
-        assert check_data_dims(kraus_data)
+        if len(self.__data) > 4**self.__nqubit:
+            raise ValueError("len(kraus_data) cannot exceed 4**nqubit.")
 
-        # check that the channel is properly normalized i.e
-        # \sum_K_i^\dagger K_i = Identity
-        assert check_data_normalization(kraus_data)
-        assert check_rank(kraus_data)
+        # Check that the channel is properly normalized, i.e., \sum_K_i^\dagger K_i = Identity.
+        data = next(iter(self.__data))
+        work = np.zeros_like(data.operator, dtype=np.complex128)
+        for data in self.__data:
+            m = data.coef * data.operator
+            work += m.conj().T @ m
+        if not np.allclose(work, np.eye(2**self.__nqubit)):
+            raise ValueError("The specified channel is not normalized.")
 
-        self.nqubit = int(np.log2(kraus_data[0]["operator"].shape[0]))
-        self.kraus_ops = kraus_data
+    @typing.overload
+    def __getitem__(self, index: SupportsIndex, /) -> KrausData: ...
 
-        # np.asarray(data, dtype=np.complex128)
-        # number of Kraus operators in the Channel
+    @typing.overload
+    def __getitem__(self, index: slice, /) -> list[KrausData]: ...
 
-        self.size = len(kraus_data)
+    def __getitem__(self, index: SupportsIndex | slice, /) -> KrausData | list[KrausData]:
+        """Return the Kraus operator at the given index."""
+        return copy.deepcopy(self.__data[index])
```

Comment:
Do we need a `deepcopy` here? We could just return a read-only view.

Reply:
>  just return a read-only view

How can we do this?
I don't know any method essentially faster than `copy.deepcopy` .

Reply:
`KrausData` _is_ a read-only view, isn't it?

Reply:
Not completely read-only as `KrausData` contains mutable classes.

Reply:
Anyway, unlike `deepcopy` in `__init__` that is always necessary, we could be optimistic here...

Reply:
Fixed in 7ad9425f4933b0c2d963f060cd432e934dba1333 .

Reply:
> Not completely read-only as `KrausData` contains mutable classes.

Indeed, but the mutability does not seem to be publicly exposed, isn't it? I mean, mutable data is private and accessors make the interface immutable, as far as I understand.